### PR TITLE
Chapter comments

### LINKED ISSUE #14: Comic/Chapters comments
add comments to both chapters and comics

chapter and comic comments should have different tables to differentiate between them

comments will have children comments and parent comments so this can be done by adding a `comment_parents_children` table that connects parents and children 

models:

```rust
// POST /api/v1/comics/chapters/:chapter_id/comments
struct CreateChapterComment {
    content: String,
}
```

```rust
// POST /api/v1/comics/:comic_id/comments
struct CreateComicComment {
    content: String,
}
```
------
**UPDATE:**
since having to render comments recursively with it's child comments, it seems a bit unnecessary and over complicated for what is important, to make it simple, it could be simplified to only have one level of comments replies like so:

1- table for comments on comics/chapters
2- table for replies on comments

with relations:
one comic to many comments
one comment to many replies

and that's it :)

OR only render 1 level of comment replies then require the user to enter the thread to get to the children of the reply

but it still feels a bit unnecessary for an indie comic platform to have so much depth to the comments and replies

but we will see what the future holds :)

------
**UPDATE2:**
I figured out a way to send the necessary data to the client, by inlining diesel's `belonging_to()` function:

```rust
    let comment_mappings_by_parents = comment_mappings.clone().grouped_by(&comments); // 1 parent -> children

    let id_indices: HashMap<_, _> = comments
        .iter()
        .enumerate()
        .map(|(i, u)| (u.id, i))
        .collect();

    let mut comment_mappings_by_children = comments.iter().map(|_| Vec::new()).collect::<Vec<_>>();

    for child in comment_mappings {
        comment_mappings_by_children[id_indices[&child.child_comment_id]].push(child);
    }
```

and adding the parent id to the comment struct:

```rust
pub struct ComicCommentResponse {
    pub id: Uuid,
    pub content: String,
    pub user: UserResponseBrief,
    pub parent_comment: Option<Uuid>, // this part
    pub child_comments: Option<Vec<Uuid>>,
}
```
now in the frontend, I can first get the top level comments aka the ones with no parents, then recursively add the children to them. [frontend implementation](https://github.com/BKSalman/musawarah/issues/22)

------

### Todos:
- [x] create chapter_comments migration
- [x] create chapter_comments_parents_children migration
- [x] create comic_comments migration
- [x] create comic_comments_parents_children migration
- [x] handler for creating comments that recieves `CreateComment` as payload and gets current user with `AuthExtractor`
- [x] save comment to DB
- [x] link comments to parent comments
- [x] handler for deleting comments
- [x] handler to get chapter comments
- [x] handler to get comic comments

### PR DESCRIPTION
**Pull request by BKSalman**
### Todos:
- [x] add chapter comments migrations
- [x] add chapter comments mapping migrations
- [x] get comments handler
- [x] create comment handler
- [x] delete comment handler
- [x] frontend
    - [x] get comments
    - [x] display comments
    - [x] a way to send a comment

### CODE REVIEW

**Comment by BKSalman** [comment]
code is kinda shitty, if it's worth, maybe give me some points on how to improve it

File: client/src/lib/components/Comment.svelte [hunk 1]
```diff
@@ -5,12 +5,17 @@
     import Fa from "svelte-fa";
     import { currentUser } from "../../routes/stores";
     import { fly } from "svelte/transition";
+    import type { ChapterCommentResponse } from "bindings/ChapterCommentResponse";
 
-    export let comment: ComicCommentResponse;
+    export let comment: ComicCommentResponse | ChapterCommentResponse;
     export let indent = 0;
-    let openChildren = false;
+    let openChildren = comment.parent_comment == null;
```
Comment (mohad12211) [inline client/src/lib/components/Comment.svelte:12]: problem with this is that if you add a reply under a reply, it's collapsed by default

**Review by mohad12211 — CHANGES_REQUESTED** [review]

File: client/src/routes/[username]/[comic_slug]/[chapter_number]/+page.svelte [hunk 2]
```diff
@@ -1,30 +1,89 @@
...
 
     export let data: PageServerData;
 
-    const { chapter } = data;
+    let { chapter, comments } = data;
+
+    async function sendComment(
+        e: SubmitEvent,
+        parent_comment_id: string | null,
+        comic_id: string
```
Comment (mohad12211) [inline client/src/routes/[username]/[comic_slug]/[chapter_number]/+page.svelte]:
```suggestion
        chapter_id: string
```

File: client/src/routes/[username]/[comic_slug]/[chapter_number]/+page.svelte [hunk 3]
```diff
@@ -1,30 +1,89 @@
...
+        const form = new FormData(e.target as HTMLFormElement);
+
+        const comment = form.get("comment");
+
+        if (comment?.toString() == null || comment?.toString().length < 1) {
+            return;
+        }
+
+        const res = await fetch(
+            `http://localhost:6060/api/v1/comics/${comic_id}/comments`,
```
Comment (mohad12211) [inline client/src/routes/[username]/[comic_slug]/[chapter_number]/+page.svelte]:
```suggestion
            `http://localhost:6060/api/v1/comics/chapters/${chapter_id}/comments`,
```

**Review by mohad12211 — CHANGES_REQUESTED** [review]
copy paste mistakes

File: client/src/routes/[username]/[comic_slug]/[chapter_number]/+page.svelte (hunk 3 above)
Comment (BKSalman) [reply client/src/routes/[username]/[comic_slug]/[chapter_number]/+page.svelte]: wtf, how, it shouldn't have worked with this bug

Comment (mohad12211) [reply client/src/routes/[username]/[comic_slug]/[chapter_number]/+page.svelte]: yeah it didn't work when I tried it

Comment (BKSalman) [reply client/src/routes/[username]/[comic_slug]/[chapter_number]/+page.svelte]: ah, yeah, I didn't check the correctness of this file, mb

File: client/src/lib/components/Comment.svelte (hunk 1 above)
Comment (BKSalman) [reply client/src/lib/components/Comment.svelte:12]: I made it like this to not clutter the screen with nested replies, so only top level replies (not to be confused with comments) will be open.

also when sending a reply I will make it open the reply replied to

Comment (mohad12211) [reply client/src/lib/components/Comment.svelte:12]: yeah good idea

File: src/comics/chapters/chapter_comments/routes.rs [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+        (status = StatusCode::BAD_REQUEST, description = "Invalid Comment ID", body = ErrorResponse),
+        (status = StatusCode::INTERNAL_SERVER_ERROR, description = "Something went wrong", body = ErrorResponse),
+    ),
+    tag = "Chapter Comments API"
+)]
+#[axum::debug_handler(state = AppState)]
+pub async fn delete_comment(
+    auth: AuthExtractor<{ UserRole::User as u32 }>,
+    State(state): State<Arc<InnerAppState>>,
+    Path(comment_id): Path<Uuid>,
```
Comment (mohad12211) [inline src/comics/chapters/chapter_comments/routes.rs:198]: this is picking the chapter id.
the route should be `/api/v1/comics/chapters/comics/:comment:id`

Comment (BKSalman) [reply src/comics/chapters/chapter_comments/routes.rs:198]: done (it's `/api/v1/comics/chapters/comments/:comment_id` btw)

**Review by mohad12211 — APPROVED** [review]
LGTM LGTM

**Pull request merged**